File: 19.删除链表的倒数第-n-个结点.py
#
# @lc app=leetcode.cn id=19 lang=python
#
# [19] 删除链表的倒数第 N 个结点
#
# 解法1(T86% S84%)：首先添加头指针，一次遍历链表，遍历过程中计数并把pre指针存下来，全部遍历完获取倒数第n个的pre指针，最后进行一次删除

# @lc code=start
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def removeNthFromEnd(self, head, n):
        """
        :type head: ListNode
        :type n: int
        :rtype: ListNode
        """
        self.next = head
        preNode = [self]
        length = 1
        while head:
            preNode.append(head)
            length += 1
            head = head.next
        pre = preNode[length-n-1]
        pre.next = pre.next.next
        
        return self.next
# @lc code=end

File: javascript/19.删除链表的倒数第-n-个结点.js
/*
 * @lc app=leetcode.cn id=19 lang=javascript
 *
 * [19] 删除链表的倒数第 N 个结点
 * 
 * 快慢指针
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function(head, n) {
    if (!head.next && n==1) { return head.next; }

    var fast = head, slow = head;
    for (var i=0; i<n; i++) {
        fast = fast.next;
    }
    // 如果此时快指针已经出界，则删除的就是第一个节点
    if (!fast) { return head.next; }

    while (fast && fast.next) {
        fast = fast.next;
        slow = slow.next;
    }
    slow.next = slow.next.next;
    return head;
};
// @lc code=end

